// check-image-type/dist/check-image-type.esm.js
function _extends() {
  _extends = Object.assign || function (target) {
    for (var i = 1; i < arguments.length; i++) {
      var source = arguments[i];

      for (var key in source) {
        if (Object.prototype.hasOwnProperty.call(source, key)) {
          target[key] = source[key];
        }
      }
    }

    return target;
  };

  return _extends.apply(this, arguments);
}

function _unsupportedIterableToArray(o, minLen) {
  if (!o) return;
  if (typeof o === "string") return _arrayLikeToArray(o, minLen);
  var n = Object.prototype.toString.call(o).slice(8, -1);
  if (n === "Object" && o.constructor) n = o.constructor.name;
  if (n === "Map" || n === "Set") return Array.from(o);
  if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen);
}

function _arrayLikeToArray(arr, len) {
  if (len == null || len > arr.length) len = arr.length;

  for (var i = 0, arr2 = new Array(len); i < len; i++) arr2[i] = arr[i];

  return arr2;
}

function _createForOfIteratorHelperLoose(o, allowArrayLike) {
  var it = typeof Symbol !== "undefined" && o[Symbol.iterator] || o["@@iterator"];
  if (it) return (it = it.call(o)).next.bind(it);

  if (Array.isArray(o) || (it = _unsupportedIterableToArray(o)) || allowArrayLike && o && typeof o.length === "number") {
    if (it) o = it;
    var i = 0;
    return function () {
      if (i >= o.length) return {
        done: true
      };
      return {
        done: false,
        value: o[i++]
      };
    };
  }

  throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method.");
}

var WEBP = {
  ext: "webp",
  mime: "image/webp"
};
var JPG = {
  ext: "jpg",
  mime: "image/jpeg"
};
var PNG = {
  ext: "png",
  mime: "image/png"
};
var BMP = {
  ext: "bmp",
  mime: "image/bmp"
};
var GIF = {
  ext: "gif",
  mime: "image/gif"
};
var AVIF = {
  ext: "avif",
  mime: "image/avif"
};
var HEIC = {
  ext: "heic",
  mime: "image/heic"
};
var BPG = {
  ext: "bpg",
  mime: "image/bpg"
};
var ICO = {
  ext: "ico",
  mime: "image/ico"
};
var PSD = {
  ext: "psd",
  mime: "image/vnd.adobe.photoshop"
};
var JXR = {
  ext: "jxr",
  mime: "image/vnd.ms-photo"
};
var FLIF = {
  ext: "flif",
  mime: "image/flif"
};
var ICNS = {
  ext: "icns",
  mime: "image/icns"
};
var ORF = {
  ext: "orf",
  mime: "image/x-olympus-orf"
};
var XCF = {
  ext: "xcf",
  mime: "image/x-xcf"
};
var RW2 = {
  ext: "rw2",
  mime: "image/x-panasonic-rw2"
};
var KTX = {
  ext: "ktx",
  mime: "image/ktx"
};
var JXL = {
  ext: "jxl",
  mime: "image/jxl"
};
var CUR = {
  ext: "cur",
  mime: "image/x-icon"
};
var RAF = {
  ext: "raf",
  mime: "image/x-fujifilm-raf"
};

var defaultOption = {
  offset: 0
};

var check = function check(buffer, data, option) {
  var _defaultOption$option = _extends({}, defaultOption, option),
      offset = _defaultOption$option.offset;

  return data.every(function (v, k) {
    return buffer[k + offset] === v;
  });
};

var checkString = function checkString(buffer, str, option) {
  return check(buffer, Array.from(str).map(function (i) {
    return i.charCodeAt(0);
  }), option);
};

var isBMP = function isBMP(buffer) {
  if (!buffer || buffer.length < 2) {
    return false;
  }

  return check(buffer, [0x42, 0x4d]);
};
var isGIF = function isGIF(buffer) {
  if (!buffer || buffer.length < 3) {
    return false;
  }

  return check(buffer, [0x47, 0x49, 0x46]);
};
var isJPG = function isJPG(buffer) {
  if (!buffer || buffer.length < 3) {
    return false;
  }

  return check(buffer, [0xff, 0xd8, 0xff]);
};
var isJXR = function isJXR(buffer) {
  if (!buffer || buffer.length < 3) {
    return false;
  }

  return check(buffer, [0x49, 0x49, 0xbc]);
};
var isFLIF = function isFLIF(buffer) {
  if (!buffer || buffer.length < 4) {
    return false;
  }

  return checkString(buffer, "FLIF");
};
var isXCF = function isXCF(buffer) {
  if (!buffer || buffer.length < 12) {
    return false;
  }

  return checkString(buffer, "gimp xcf ");
};
var isORF = function isORF(buffer) {
  if (!buffer || buffer.length < 12) {
    return false;
  }

  return check(buffer, [0x49, 0x49, 0x52, 0x4f, 0x08, 0x00, 0x00, 0x00, 0x18]);
};
var isRW2 = function isRW2(buffer) {
  if (!buffer || buffer.length < 12) {
    return false;
  }

  return check(buffer, [0x49, 0x49, 0x55, 0x00, 0x18, 0x00, 0x00, 0x00, 0x88, 0xe7, 0x74, 0xd8]);
};
var isPNG = function isPNG(buffer) {
  if (!buffer || buffer.length < 8) {
    return false;
  }

  return check(buffer, [0x89, 0x50, 0x4e, 0x47, 0x0d, 0x0a, 0x1a, 0x0a]);
};
var isWEBP = function isWEBP(buffer) {
  if (!buffer || buffer.length < 12) {
    return false;
  }

  return checkString(buffer, "WEBP", {
    offset: 8
  });
};
var isICNS = function isICNS(buffer) {
  if (!buffer || buffer.length < 12) {
    return false;
  }

  return checkString(buffer, "icns");
};
var isAVIF = function isAVIF(buffer) {
  if (!buffer || buffer.length < 12) {
    return false;
  }

  return checkString(buffer, "ftypavif", {
    offset: 4
  });
};
var isHEIC = function isHEIC(buffer) {
  if (!buffer || buffer.length < 12) {
    return false;
  }

  return checkString(buffer, "ftyp", {
    offset: 4
  }) && (checkString(buffer, "mif1", {
    offset: 8
  }) || checkString(buffer, "msf1", {
    offset: 8
  }) || checkString(buffer, "heic", {
    offset: 8
  }) || checkString(buffer, "heix", {
    offset: 8
  }) || checkString(buffer, "hevc", {
    offset: 8
  }) || checkString(buffer, "hevx", {
    offset: 8
  }));
};
var isBPG = function isBPG(buffer) {
  if (!buffer || buffer.length < 12) {
    return false;
  }

  return check(buffer, [0x42, 0x50, 0x47, 0xfb]);
};
var isICO = function isICO(buffer) {
  if (!buffer || buffer.length < 12) {
    return false;
  }

  return check(buffer, [0x00, 0x00, 0x01, 0x00]);
};
var isPSD = function isPSD(buffer) {
  if (!buffer || buffer.length < 12) {
    return false;
  }

  return checkString(buffer, "8BPS");
};
var isKTX = function isKTX(buffer) {
  if (!buffer || buffer.length < 12) {
    return false;
  }

  return check(buffer, [0xab, 0x4b, 0x54, 0x58, 0x20, 0x31, 0x31, 0xbb, 0x0d, 0x0a, 0x1a, 0x0a]);
};
var isJXL = function isJXL(buffer) {
  if (!buffer || buffer.length < 12) {
    return false;
  }

  return check(buffer, [0xff, 0x0a]) || check(buffer, [0x00, 0x00, 0x00, 0x0c, 0x4a, 0x58, 0x4c, 0x20, 0x0d, 0x0a, 0x87, 0x0a]);
};
var isCUR = function isCUR(buffer) {
  if (!buffer || buffer.length < 12) {
    return false;
  }

  return check(buffer, [0x00, 0x00, 0x02, 0x00]);
};
var isRAF = function isRAF(buffer) {
  if (!buffer || buffer.length < 12) {
    return false;
  }

  return checkString(buffer, "FUJIFILMCCD-RAW");
};
var pipe = [[isJPG, JPG], [isPNG, PNG], [isGIF, GIF], [isWEBP, WEBP], [isBMP, BMP], [isAVIF, AVIF], [isHEIC, HEIC], [isICO, ICO], [isBPG, BPG], [isPSD, PSD], [isJXR, JXR], [isFLIF, FLIF], [isICNS, ICNS], [isXCF, XCF], [isRW2, RW2], [isKTX, KTX], [isJXL, JXL], [isCUR, CUR], [isORF, ORF], [isRAF, RAF]];
var checkImage = function checkImage(buffer) {
  for (var _iterator = _createForOfIteratorHelperLoose(pipe), _step; !(_step = _iterator()).done;) {
    var _step$value = _step.value,
        fun = _step$value[0],
        type = _step$value[1];
    if (fun(buffer)) return type;
  }

  return undefined;
};

export { AVIF, BMP, BPG, CUR, FLIF, GIF, HEIC, ICNS, ICO, JPG, JXL, JXR, KTX, ORF, PNG, PSD, RAF, RW2, WEBP, XCF, checkImage, isAVIF, isBMP, isBPG, isCUR, isFLIF, isGIF, isHEIC, isICNS, isICO, isJPG, isJXL, isJXR, isKTX, isORF, isPNG, isPSD, isRAF, isRW2, isWEBP, isXCF };
